### PR TITLE
Digram

### PR DESCRIPTION
**Pull request by aalexandrasilva**
Added the following implementation:
1) Finding all digram occurrences in a graph (any two edges connected to a common node), by using three main cases:
- Common node is object in one, and subject in other: n1 e1 **n2** -> **n2** e2 n3
- Common node is object in both: n1 e1 **n2** -> n3 e2 **n2**
- Common node is subject in both: **n2** e1 n1 -> **n2** e2 n3

2) Sorting the digrams by most to least frequent and discarding the overlapping occurrences

### CODE REVIEW

File: src/main/java/org/dice_group/grp/grammar/digram/Digram.java [hunk 1]
```diff
@@ -82,4 +78,51 @@ public void setNoOfOccurences(long occurences) {
...
+		return result;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (obj == null)
+			return false;
+		if (getClass() != obj.getClass())
```
Comment (TortugaAttack) [inline src/main/java/org/dice_group/grp/grammar/digram/Digram.java]: instead you can ask if(obj instanceof Digram) 
In this case it does not matter though

File: src/main/java/org/dice_group/grp/grammar/digram/Digram.java [hunk 2]
```diff
@@ -82,4 +78,51 @@ public void setNoOfOccurences(long occurences) {
...
+			if (other.edgeLabel1 != null)
+				return false;
+		} else if (!edgeLabel1.equals(other.edgeLabel1))
+			return false;
+		if (edgeLabel2 == null) {
+			if (other.edgeLabel2 != null)
+				return false;
+		} else if (!edgeLabel2.equals(other.edgeLabel2))
+			return false;
+		if (external == null) {
```
Comment (TortugaAttack) [inline src/main/java/org/dice_group/grp/grammar/digram/Digram.java]: externals are not allowed to be empty/null according to the paper. So this case can be removed

File: src/main/java/org/dice_group/grp/grammar/digram/Digram.java [hunk 3]
```diff
@@ -82,4 +78,51 @@ public void setNoOfOccurences(long occurences) {
...
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + ((edgeLabel1 == null) ? 0 : edgeLabel1.hashCode());
+		result = prime * result + ((edgeLabel2 == null) ? 0 : edgeLabel2.hashCode());
+		result = prime * result + ((external == null) ? 0 : external.hashCode());
+		return result;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
```
Comment (TortugaAttack) [inline src/main/java/org/dice_group/grp/grammar/digram/Digram.java]: all the edgeLabelX == null can be removed, if an edgeLabel is null there is something wrong. :)

File: src/test/java/org/dice_group/grp/grammar/digram/DigramTest.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.Set;
+
+import org.apache.jena.rdf.model.Model;
+import org.apache.jena.rdf.model.ModelFactory;
+import org.junit.Test;
+
+public class DigramTest {
+
+	@Test
+	public void test() {
```
Comment (TortugaAttack) [inline src/test/java/org/dice_group/grp/grammar/digram/DigramTest.java]: should have an assertion that everything went correct.

File: src/main/java/org/dice_group/grp/grammar/digram/DigramHelper.java [hunk 5]
```diff
@@ -47,14 +68,275 @@
...
+	/**
+	 * Returns true if the triple count for a given node is positive
+	 * @param nodeString
+	 * @param diff
+	 * @param graph
+	 * @return
+	 */
+	private static boolean isNodeExternal(String nodeString, int diff, Model graph) {
+		StringBuilder query = new StringBuilder();
+		query.append("SELECT (COUNT(*)-").append(diff).append(" AS ?total) WHERE { { ")
```
Comment (TortugaAttack) [inline src/main/java/org/dice_group/grp/grammar/digram/DigramHelper.java]: Shouldnt this be an ASK query with MINUS, as we discussed?

File: src/main/java/org/dice_group/grp/grammar/digram/DigramHelper.java [hunk 6]
```diff
@@ -47,14 +68,275 @@
...
+		}
+		return false;
+	}
+	
+	/**
+	 * Returns a string encapsulated with < > in case of an URI resource and " " in case of a literal
+	 * @param node
+	 * @return
+	 */
+	private static String getNodeName(RDFNode node) {
```
Comment (TortugaAttack) [inline src/main/java/org/dice_group/grp/grammar/digram/DigramHelper.java]: There is a class called JenaNodeFormatter.format() in the HDT library (included) this does exactly that :)

File: src/main/java/org/dice_group/grp/grammar/digram/DigramHelper.java [hunk 7]
```diff
@@ -47,14 +68,275 @@
...
+		return nodeStr;
+	}
+	
+	/**
+	 * creates a sparql query for a local model
+	 * @param graph
+	 * @param sparqlQuery
+	 * @return
+	 */
+	public static QueryExecution queryModel(Model graph, String sparqlQuery) {
```
Comment (TortugaAttack) [inline src/main/java/org/dice_group/grp/grammar/digram/DigramHelper.java]: maybe queryModel, getNodeName and selectModel in its own RDFHelper class. So it is a little bit more lucid

File: src/main/java/org/dice_group/grp/grammar/digram/Digram.java (hunk 2 above)
Comment (MichaelRoeder) [reply src/main/java/org/dice_group/grp/grammar/digram/Digram.java]: Well, following the idea of defensive programming, you may simply want to throw an exception with an error message indicating that this case is not allowed :wink:

Comment (TortugaAttack) [reply src/main/java/org/dice_group/grp/grammar/digram/Digram.java]: true.

File: src/main/java/org/dice_group/grp/grammar/digram/Digram.java (hunk 3 above)
Comment (MichaelRoeder) [reply src/main/java/org/dice_group/grp/grammar/digram/Digram.java]: Again, you could also throw an exception. Or use [`Objects.requireNonNull()`](https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#requireNonNull(T,%20java.lang.String))

File: src/main/java/org/dice_group/grp/grammar/digram/DigramHelper.java [hunk 8]
```diff
@@ -47,14 +68,275 @@
...
 	
-	public static Set<DigramOccurence> findDigrams(Model graph, Digram dig) {
-		// TODO Auto-generated method stub
-		return null;
+	/**
+	 * Non overlapping occurrences, priority given to the most frequent digrams
+	 * @param occurrences
+	 * @return
+	 */
+	public static Map<Digram, Set<DigramOccurence>> findNonOverOccurrences(Set <DigramOccurence> occurrences){
```
Comment (TortugaAttack) [inline src/main/java/org/dice_group/grp/grammar/digram/DigramHelper.java]: I like the worst case O(n^2) best case O(n) greedy approach. looks good to me.
However this is heuristic, we might have to come back to check if we can get another approach implemented in the same worst case time which is not heuristic. 
Then we can compare the two approaches and check if the greedy algorithm is sufficient in most cases.

Comment (TortugaAttack) [reply src/main/java/org/dice_group/grp/grammar/digram/DigramHelper.java]: Acutally i have one written on paper, so i can do that.

File: src/main/java/org/dice_group/grp/grammar/digram/Digram.java (hunk 2 above)
Comment (aalexandrasilva) [reply src/main/java/org/dice_group/grp/grammar/digram/Digram.java]: my bad, I used the auto-generated equals by the IDE. But I changed it to incorporate your reviews :)

**Comment by aalexandrasilva** [comment]
> Pull Request looks good :)
> 
> Minor things however as stated
> Especially the one with the MINUS query and the test assertion.

I have changed the code, the .equals() method has been rectified and as we've discussed the minus query wasn't possible with Jena. 

I've also added new assertions in the test cases :)

**Comment by TortugaAttack** [comment]
Thanks :)

**Pull request merged**